### ISSUE TITLE
Lazy exit wave stack != computed exit wave stack

### ISSUE THREAD

**Issue by alouiset** [abTEM/abTEM#154]
Hi,
I want to determine how much I can increase the slice_thickness parameter for a specific multislice simulation and I ran into an issue with the abtem.stack() function while testing multislice convergence. I am using abtem version = 1.0.4 with abtem.config.set({"device": "gpu"}), abtem.config.set({"precision": "float64"}) and only using the 'finite' projection for potential calculations.
I am comparing the simulated exit waves for different slice thicknesses with a "ground truth" (ew_truth in the following), which was calculated using an overly small slice thickness.

I have the following function which returns the exit wave ensemble to be tested:
```
def slice_thickness_convergence_series(
    atoms : ase.Atoms,
    wave : abtem.Waves,
    sampling : float | tuple,
    slice_thicknesses : list | np.ndarray,
    thickness_fractions : list | np.ndarray,
    projection : str = 'finite',
    compute : bool = True
):
    exit_waves_list = []
    for slice_thickness in slice_thicknesses:
        potential = abtem.Potential(
            atoms,
            sampling=sampling,
            slice_thickness=slice_thickness,
            projection=projection
        )
        if compute:
            exit_wave = wave.multislice(potential).compute()
        else:
            exit_wave = wave.multislice(potential)
        
        exit_waves_list.append(exit_wave)

    slice_thickness_axes_metadata = abtem.core.axes.OrdinalAxis(
        label="slice thickness", values=[f"cc/{f}" for f in thickness_fractions]
    )

    return abtem.stack(
        exit_waves_list,
        axis_metadata=slice_thickness_axes_metadata,
    )
```
Then I have the following parameters to test convergence:
```
cc_bond = 1.42 # in Å
thickness_fractions = np.array([2**n for n in np.arange(1, 7)])
slice_thicknesses = cc_bond / thickness_fractions
```
These parameters are chosen so, when given to slice_thickness_convergence_series(), the last element of the resulting abtem.stack() should correspond to the ground truth I previously defined (ew_truth variable). However, I do not get the same results if I compute the arrays before or after creating the ensemble. Here is an example:
```
test_ew_compute = slice_thickness_convergence_series(
    atoms,
    probe,
    sampling = sampling,
    slice_thicknesses = slice_thicknesses,
    thickness_fractions = thickness_fractions,
    compute=True
)

test_ew_lazy = slice_thickness_convergence_series(
    atoms,
    probe,
    sampling = sampling,
    slice_thicknesses = slice_thicknesses,
    thickness_fractions = thickness_fractions,
    compute=False
)
test_ew_lazy.compute()

for i, exit_wave in enumerate(test_ew_compute):
    print('Are tested exit waves {} close ? ->'.format(i), np.allclose(test_ew_lazy[i].array, exit_wave.array))
    print('exit wave compute {} close to ground truth ? ->'.format(i), np.allclose(ew_truth.array, exit_wave.array))
    print('exit wave lazy, then compute {} close to ground truth ? ->'.format(i), np.allclose(ew_truth.array, test_ew_lazy[i].array),'\n')
```
This returns:
```
Are tested exit waves 0 close ? -> True
exit wave compute 0 close to ground truth ? -> False
exit wave lazy, then compute 0 close to ground truth ? -> False 

Are tested exit waves 1 close ? -> False
exit wave compute 1 close to ground truth ? -> False
exit wave lazy, then compute 1 close to ground truth ? -> False 

Are tested exit waves 2 close ? -> False
exit wave compute 2 close to ground truth ? -> False
exit wave lazy, then compute 2 close to ground truth ? -> False 

Are tested exit waves 3 close ? -> False
exit wave compute 3 close to ground truth ? -> False
exit wave lazy, then compute 3 close to ground truth ? -> False 

Are tested exit waves 4 close ? -> False
exit wave compute 4 close to ground truth ? -> False
exit wave lazy, then compute 4 close to ground truth ? -> False 

Are tested exit waves 5 close ? -> False
exit wave compute 5 close to ground truth ? -> True
exit wave lazy, then compute 5 close to ground truth ? -> False  
```

I was expecting all tested  exit waves [i] to be identical, or only the first ones are (i=0). Then the last ones (i=5) should be identical to ew_truth. While it is the case when slice_thickness_convergence_series(compute=True), it is not when I run slice_thickness_convergence_series(compute=False) then .compute()

I don't understand what is happening here. It looks like computing abtem.stack(abtem.Waves()).multislice() is not equivalent to abtem.Waves().multislice(). In my case, neither changing the precision nor forcing small values to zeros with np.round() solved this issue.
Do you have any idea what is happening here?
Thank you!
Antonin.

**Comment by TomaSusi** [comment]
Hi! Could you please provide the full code, including the `atoms` and the `Probe`?

**Comment by TomaSusi** [comment]
I think I found an issue.

After running `test_ew_lazy.compute()`, only the first exit-wave in the stack seems to actually be computed correctly, while the others have arrays filled with `nan`s. 

I think this looks like it could be a bug, could you have a look @jacobjma?

**Comment by alouiset** [comment]
Hi Toma,

I want to simulate 4DSTEM datasets of defective carbon nanotube structures calculated by molecular dynamics. The atoms repartition in the z direction is quite perculiar so I can not rely on actual interplanar distances to set the slice thickness parameter. Using abtem.slicing.crystal_slice_thicknesses() is working but it gives a huge number of slices so I can not use it for "reasonable" computation time. Maybe I can use this function in a better way but I couldn't figure out how. 

So I just chose the c-c bond length in graphene as a reference value for my slice thicknesses and tested different fractions. I'm trying to see how much I can increase the slice thickness without significantly affecting the precision of my simulations. I actually already ran my test by comparing the diffraction pattern intensities for a single probe position. While doing so, I encounter this issue.
Here you'll find a test notebook (just an example for real space exit waves) with the associated ase.Atoms() in the .xyz file.
[stack_lazy_ensemble_issue.zip](https://github.com/abTEM/abTEM/files/15336175/stack_lazy_ensemble_issue.zip)
In my case, all exit waves are computed. Values in the arrays are just slightly different. For example:
```
test_ew_compute[3].array[35,128] = (-2.4704542e-08+1.1070954e-08j)
test_ew_lazy[3].array[35,128] = (-2.4704907e-08+1.1073689e-08j)
```
It's not affecting my work at the moment, but I'm afraid it could become a problem in the future.
Thank you for your time,
Antonin.

**Comment by TomaSusi** [comment]
Okay, I see where you were getting the slice thickness from, I guess it makes some sense if you have bonds on the side walls also parallel to the beam direction. Usually we just pick 1 Å as a coarse slice regardless of the atoms object, and 0.5 Å is typically sufficient for practical purposes. I see also now that for testing with the view of doing 4D-STEM simulations, a probe with single position is a reasonable starting point.

Thanks for the notebook and structure file! I have a few comments:
- it's not relevant here, but note that even though your atoms object doesn't have periodic boundary conditions in all directions, our potential will always be periodic.
- a sampling of 0.08875 Å will definitely NOT be fully converged; typically we use 0.05 or 0.04 Å, but you might need to go even lower for high numerical precision.

I could now reproduce the discrepancy you see, though it's curious that with a `Probe` I don't get the `nan`s I got with a `PlaneWave`!

I ran the same calculations, and I notice that while the direct compute seems to give the same numerical values every time, the lazy stack computed afterwards seems to return slightly different values each time. I wonder if this is expected behavior, @jacobjma, or an indication of an issue?

I do note that the differences are quite minor and probably should be negligible for practical purposes. Nonetheless, I agree that it is odd that computing a stack would give different results than stacking computed exit waves.

**Comment by alouiset** [comment]
Thanks for your comments! To give you some feedback:

- I actually run my simulations on larger supercells to avoid periodicity artifacts. Also to get a smaller `DiffractionPatterns` pixel size.
- I agree that 0.08875 Å is not sufficient for high precision, especially at high spatial frequencies, but I only need to compute `DiffractionPatterns` up to small scattering angles, typically 60 mrad. I compared with a sampling down to 0.02 Å and relative errors stay quite small (few percentages). However they get super high super quick if I compute `DiffractionPatterns` up to larger scattering angles. In my case, this seems to be a good compromise between precision and faster computation.
- Yes I agree. my test showed very small differences between slice_thickness = 0.04 and 0.08 Å. Looks like I can actually work with 0.355 Å (cc_bond/4). At low scattering angles (<40 mrad), slice_thickness = 0.71 Å (cc_bond/2) does not even look that bad.

Yes I agree, value differences are small and do not have a significant impact on the final results, for a physicist perspective at least. I guess it's more concerning from a numerical calculation point of view, especially for precision tests.

**Comment by TomaSusi** [comment]
Excellent, looks like you know what you're doing, all that sounds very reasonable for your specific use case. We'll look into to the numerical issue, thanks for flagging this!

**Comment by jacobjma** [comment]
Hi @TomaSusi @alouiset. Thank you for the detailed issue and discussion. The good news is that I am pretty sure it is already solved in `1.0.5`, which was out a couple of weeks ago.

float64 precision was first introduced in `1.0.4`, and I found a couple of issues relating to that. I am not sure what about stacking triggered this. 

For future reference, this is a minimal reproducible example triggering the issue in `1.0.4`, but not in `1.0.5`:
```python
import abtem
import ase
import numpy as np

pw = abtem.PlaneWave(energy=80e3, sampling=0.2)

atoms = ase.build.bulk("Si", cubic=True)

potentials = [abtem.Potential(atoms, slice_thickness=t) for t in (1, 2)]

exit_wave_compute = [pw.multislice(potential).compute() for potential in potentials]
exit_wave_compute_stack = abtem.stack(exit_wave_compute)

exit_waves = [pw.multislice(potential) for potential in potentials]
exit_wave_stack_compute = abtem.stack(exit_waves).compute()

assert np.allclose(exit_wave_stack_compute.array, exit_wave_compute_stack.array)
```

**Comment by alouiset** [comment]
Hi @jacobjma,
I just tested your script on my server and I got the expected result: version = 1.0.4 -> AssertionError, version = 1.0.5 -> no error. I'm switching to version 1.0.5. Thank you very much!
